### ISSUE TITLE
Ensure more accurate `TData` TypeScript types for `useSuspenseQuery`

### ISSUE THREAD

**Issue by jerelmiller** [apollographql/apollo-client#10713]
Currently `useSuspenseQuery` sets the returned `data` property to type `TData` (a generic). This works well with the default behavior since the hook will throw errors when encountered.

There are however a couple cases that will cause `data` to remain `undefined`.

1. Setting `errorPolicy` to anything other than `none` (the default). If an error is encountered with `all` or `ignore`, the error is not thrown and is expected to be handled locally, but `data` might not be set.
2. Setting `skip` to `true` (#10532). Once this is supported, `skip` will not load any data, so `data` is expected to be `undefined` until `skip` is `false`.

Additionally, setting `returnPartialData` may result in some of missing data. In this case, we should set `data` to `DeepPartial<TData>` (note `DeepPartial` is not a part of TypeScript: we'll need a helper) to better communicate that the `data` type might not contain all the required values.

In short, our types should reflect the following:
* `errorPolicy: 'none'` -> `data: TData`
* `errorPolicy: 'ignore'` -> `data: TData | undefined`
* `errorPolicy: 'all'` -> `data: TData | undefined`
* `skip: false` -> `data: TData`
* `skip: true` -> `data: TData | undefined`
* `returnPartialData: true` -> `data: DeepPartial<TData>`
* `returnPartialData: false` -> `data: TData`

**Comment by jerelmiller** [comment]
Fixed in https://github.com/apollographql/apollo-client/pull/10766
